### PR TITLE
Add getAMI func to get AMI based on region

### PR DESCRIPTION
**Pull request by ArangoGutierrez**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,7 +1,8 @@
 module github.com/NVIDIA/holodeck
 
-go 1.21.5
-toolchain go1.22.2
+go 1.22.0
+
+toolchain go1.22.3
```
Comment (elezar) [inline go.mod:5]: Nit. Probably not part of this PR.

File: pkg/provider/aws/create.go [hunk 2]
```diff
@@ -281,6 +281,16 @@ func (a *Client) createEC2Instance(cache *AWS) error {
 	a.log.Wg.Add(1)
 	go a.log.Loading("Creating EC2 instance")
 
+	// Check if the image is provided, if not get the latest image
+	if a.Spec.Image.ImageId == nil {
```
Comment (elezar) [inline pkg/provider/aws/create.go:285]: Should this be conditional? Can't we perform this check in `a.getAMI()`?

File: pkg/provider/aws/image.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	CreationDate string
+}
+
+type ByCreationDate []ImageInfo
+
+func (a ByCreationDate) Len() int           { return len(a) }
+func (a ByCreationDate) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a ByCreationDate) Less(i, j int) bool { return a[i].CreationDate > a[j].CreationDate }
+
+func (a *Client) getAMI() (string, error) {
```
Comment (elezar) [inline pkg/provider/aws/image.go:39]: Should this be defined where the `Client` type is defined?

File: pkg/provider/aws/image.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	filterName := "name"
+	filterArchitecture := "architecture"
+	filterOwner := "owner-id"
+	awsOwner := "099720109477" // Canonical
+
+	out, err := a.ec2.DescribeImages(context.TODO(), &ec2.DescribeImagesInput{
+		Filters: []types.Filter{
+			{
+				Name:   &filterName,
+				Values: []string{"ubuntu/images/hvm-ssd/ubuntu-jammy-22.04-amd64-server-2024*"},
```
Comment (elezar) [inline pkg/provider/aws/image.go:52]: Should this string not depend on the architecture?

File: pkg/provider/aws/image.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+func (a ByCreationDate) Less(i, j int) bool { return a[i].CreationDate > a[j].CreationDate }
+
+func (a *Client) getAMI() (string, error) {
+	var ami string
+
+	// Work around for AWS types
+	filterName := "name"
+	filterArchitecture := "architecture"
+	filterOwner := "owner-id"
+	awsOwner := "099720109477" // Canonical
```
Comment (elezar) [inline pkg/provider/aws/image.go:46]: Since these are probably the properties that we would want to change going forward, we probably want a new type that HAS A client. These can be set at construction to find the required image.

File: pkg/provider/aws/image.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+	out, err := a.ec2.DescribeImages(context.TODO(), &ec2.DescribeImagesInput{
```
Comment (elezar) [inline pkg/provider/aws/image.go:48]: Don't we expect this to also paginate the results? Why are we not handling the next token like we do above?

File: pkg/provider/aws/image.go (hunk 4 above)
Comment (ArangoGutierrez) [reply pkg/provider/aws/image.go:52]: good catch

File: pkg/provider/aws/image.go (hunk 5 above)
Comment (ArangoGutierrez) [reply pkg/provider/aws/image.go:46]: you mean, creating a NEW type on this file `image` to handle this `getAMI` func and related upcoming functions. and this type to share the `ec2` client?

Comment (ArangoGutierrez) [reply pkg/provider/aws/image.go]: We now check if the user defined a specific owner id, if not we pass the default ones

Comment (elezar) [reply pkg/provider/aws/image.go]: To be clear, this is probably only required when we start adding different conditions here.

Comment (ArangoGutierrez) [reply pkg/provider/aws/image.go]: Yup, this is more a followup, right now we are hard coded to Ubuntu, once we start opening up to RPM based OS's this would require more logic. but as it stands now, I have tested x86 and arm, and it returns a valid image AMI id

**Comment by ArangoGutierrez** [comment]
@elezar PTAL, all comments have been addressed

File: pkg/provider/aws/create.go [hunk 7]
```diff
@@ -281,6 +281,13 @@ func (a *Client) createEC2Instance(cache *AWS) error {
 	a.log.Wg.Add(1)
 	go a.log.Loading("Creating EC2 instance")
 
+	// Check if the image is provided, if not get the latest image
+	err := a.getAMI()
```
[the remark this answers is not in the document]
Comment (elezar) [reply pkg/provider/aws/create.go:285]: Alternatively call it `initAMI` or `setAMI`?

File: pkg/provider/aws/dryrun.go [hunk 8]
```diff
@@ -51,32 +51,12 @@ func (a *Client) checkInstanceTypes() error {
...
-				return nil
-			}
-		}
-
-		if resp.NextToken != nil {
-			nextToken = resp.NextToken
-		} else {
-			break
-		}
+	// Check if the given image is supported in the region
```
Comment (elezar) [inline pkg/provider/aws/dryrun.go:54]: Question: Why is this in the `dryrun` file?

File: pkg/provider/aws/image.go [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+	var filterArchitectureValue []string
+
+	if a.Spec.Instance.Image.Architecture != "" {
+		switch a.Spec.Instance.Image.Architecture {
+		case "x86_64":
+			filterArchitectureValue = []string{"x86_64", "amd64"}
+		case "arm64":
+			filterArchitectureValue = []string{"arm64"}
+		case "amd64":
+			filterArchitectureValue = []string{"x86_64", "amd64"}
```
Comment (elezar) [inline pkg/provider/aws/image.go:62]:
```suggestion
		case "x86_64", "amd64":
			filterArchitectureValue = []string{"x86_64", "amd64"}
		case "arm64":
			filterArchitectureValue = []string{"arm64"}
```

File: pkg/provider/aws/image.go [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+	filterOwner := "owner-id"
+
+	var filterNameValue []string
+	var filterArchitectureValue []string
+
+	if a.Spec.Instance.Image.Architecture != "" {
+		switch a.Spec.Instance.Image.Architecture {
+		case "x86_64":
+			filterArchitectureValue = []string{"x86_64", "amd64"}
+		case "arm64":
```
Comment (elezar) [inline pkg/provider/aws/image.go:59]: Is `aarch64` a valid architecture?

File: pkg/provider/aws/image.go [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+func (a ByCreationDate) Less(i, j int) bool { return a[i].CreationDate > a[j].CreationDate }
+
+func (a *Client) getAMI() error {
+	// If the image ID is already set by the user, return
+	if a.Spec.Image.ImageId != nil {
+		return nil
+	}
+
+	// Work around for AWS types
+	filterName := "name"
```
Comment (elezar) [inline pkg/provider/aws/image.go:48]: What is the workaround?

File: pkg/provider/aws/image.go [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+
+	var nextToken *string
+	var images []ImageInfo
+
+	for {
+		out, err := a.ec2.DescribeImages(context.TODO(), &ec2.DescribeImagesInput{
+			NextToken: nextToken,
+			Filters: []types.Filter{
+				{
+					Name:   &filterName,
```
Comment (elezar) [inline pkg/provider/aws/image.go:87]: Didn't you use `aws.String()` in another PR?

File: pkg/provider/aws/image.go [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+	CreationDate string
+}
+
+type ByCreationDate []ImageInfo
+
+func (a ByCreationDate) Len() int           { return len(a) }
+func (a ByCreationDate) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a ByCreationDate) Less(i, j int) bool { return a[i].CreationDate > a[j].CreationDate }
+
+func (a *Client) getAMI() error {
```
Comment (elezar) [inline pkg/provider/aws/image.go:41]: As I mentioned, should this be `resolve` or `set` or something other than `get`?

File: pkg/provider/aws/image.go [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+			return fmt.Errorf("invalid architecture %s", a.Spec.Instance.Image.Architecture)
+		}
+	}
+
+	// Default to the official Ubuntu images in the AWS Marketplace
+	// TODO: Add support for other image OS types
+	awsOwner := []string{"099720109477", "679593333241"}
+	if a.Spec.Instance.Image.OwnerId != nil {
+		awsOwner = []string{*a.Spec.Instance.Image.OwnerId}
+	}
```
Comment (elezar) [inline pkg/provider/aws/image.go:73]: Let's move this to before we start processing the architecture so that the archecture filter logic is together.

File: pkg/provider/aws/image.go [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+
+	for _, arch := range filterArchitectureValue {
+		filterNameValue = append(filterNameValue, fmt.Sprintf("ubuntu/images/hvm-ssd/ubuntu-jammy-22.04-%s-server-%s*", arch, getCurrentYear()))
+	}
+
+	var nextToken *string
+	var images []ImageInfo
+
```
Comment (elezar) [inline pkg/provider/aws/image.go:81]: nit:
```suggestion
	var images []ImageInfo
	var nextToken *string
```

File: pkg/provider/aws/image.go [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+	sort.Sort(ByCreationDate(images))
+	a.Spec.Image.ImageId = &images[0].ImageID
+
+	return nil
+}
+
+func (a *Client) imageIdSupported() error {
+	var nextToken *string
+
+	for {
```
Comment (elezar) [inline pkg/provider/aws/image.go:134]: do we need to factor thsi out into a `describeImages` helper since we do essentially the same thing with different input in two places.

File: pkg/provider/aws/image.go [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+	if len(images) == 0 {
+		return fmt.Errorf("no images found")
+	}
+	sort.Sort(ByCreationDate(images))
+	a.Spec.Image.ImageId = &images[0].ImageID
+
+	return nil
+}
+
+func (a *Client) imageIdSupported() error {
```
Comment (elezar) [inline pkg/provider/aws/image.go:131]: since we return an error should this be `assertImageIdSupported`?

File: pkg/provider/aws/image.go [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+			break
+		}
+	}
+
+	return fmt.Errorf("image %s is not supported in the current region %s", *a.Spec.Instance.Image.ImageId, a.Spec.Instance.Region)
+}
+
+func getCurrentYear() string {
+	now := time.Now()
+	return strconv.Itoa(now.Year())
```
Comment (elezar) [inline pkg/provider/aws/image.go:165]: return time.Now().Format("2006")

File: pkg/provider/aws/image.go [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+
+	// Default to the official Ubuntu images in the AWS Marketplace
+	// TODO: Add support for other image OS types
+	awsOwner := []string{"099720109477", "679593333241"}
+	if a.Spec.Instance.Image.OwnerId != nil {
+		awsOwner = []string{*a.Spec.Instance.Image.OwnerId}
+	}
+
+	for _, arch := range filterArchitectureValue {
+		filterNameValue = append(filterNameValue, fmt.Sprintf("ubuntu/images/hvm-ssd/ubuntu-jammy-22.04-%s-server-%s*", arch, getCurrentYear()))
```
Comment (elezar) [inline pkg/provider/aws/image.go:76]: Is using the current year brittle meaning that this tooliing won't be usable on new year's day?

File: holodeck.yml [hunk 20]
```diff
@@ -0,0 +1,29 @@
+apiVersion: holodeck.nvidia.com/v1alpha1
+kind: Environment
+metadata:
+  name: eduardoa-devel
+  description: "eduardoa devel"
+spec:
+  provider: aws
+  auth:
+    keyName: eduardoa
+    privateKey: /Users/eduardoa/.ssh/eduardoa.pem
```
[the remark this answers is not in the document]
Comment (ArangoGutierrez) [reply holodeck.yml:10]: sorry, will remove

File: pkg/provider/aws/create.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ArangoGutierrez) [reply pkg/provider/aws/create.go:285]: Good suggestions, will take `setAMI`

File: pkg/provider/aws/dryrun.go (hunk 8 above)
Comment (ArangoGutierrez) [reply pkg/provider/aws/dryrun.go:54]: Good catch, will move

File: pkg/provider/aws/image.go (hunk 11 above)
Comment (ArangoGutierrez) [reply pkg/provider/aws/image.go:48]: I forgot to use `aws.String` will address

File: pkg/provider/aws/image.go (hunk 13 above)
Comment (ArangoGutierrez) [reply pkg/provider/aws/image.go:41]: `setAMI` will be

File: pkg/provider/aws/image.go (hunk 14 above)
Comment (ArangoGutierrez) [reply pkg/provider/aws/image.go:73]: addressed

File: pkg/provider/aws/image.go (hunk 15 above)
Comment (ArangoGutierrez) [reply pkg/provider/aws/image.go:81]: thanks

File: pkg/provider/aws/image.go (hunk 19 above)
Comment (ArangoGutierrez) [reply pkg/provider/aws/image.go:76]: ummm Y2K for sure.
how about n + n-1 ?

Comment (ArangoGutierrez) [reply pkg/provider/aws/image.go:76]: how dumb of me
```go
"ubuntu/images/hvm-ssd/ubuntu-jammy-22.04-%s-server-20*"
```
this will cover us until 2099, and by 2100 we simply file a PR

File: pkg/provider/aws/image.go (hunk 18 above)
Comment (ArangoGutierrez) [reply pkg/provider/aws/image.go]: not needed under new implementation

File: pkg/provider/aws/image.go (hunk 16 above)
Comment (ArangoGutierrez) [reply pkg/provider/aws/image.go:127]: done, we now have a `func describeImages`

**Comment by ArangoGutierrez** [comment]
@elezar PTAL, all comments have been addressed

File: pkg/provider/aws/create.go [hunk 21]
```diff
@@ -281,6 +281,13 @@ func (a *Client) createEC2Instance(cache *AWS) error {
 	a.log.Wg.Add(1)
 	go a.log.Loading("Creating EC2 instance")
 
+	// Check if the image is provided, if not get the latest image
+	err := a.setAMI()
+	if err != nil {
+		a.fail()
+		return fmt.Errorf("error getting AMI: %v", err)
```
Comment (tariq1890) [inline pkg/provider/aws/create.go:288]:
```suggestion
		return fmt.Errorf("error getting AMI: %w", err)
```

File: pkg/provider/aws/dryrun.go [hunk 22]
```diff
@@ -92,11 +63,11 @@ func (a *Client) DryRun() error {
 
 	// Check if the desired image is supported in the region
 	a.log.Wg.Add(1)
-	go a.log.Loading("Checking if image %s is supported in region %s", *a.Spec.Instance.Image.ImageId, a.Spec.Instance.Region)
+	go a.log.Loading("Checking for supported image in region %s", a.Spec.Instance.Region)
 	err = a.checkImages()
 	if err != nil {
 		a.fail()
-		return fmt.Errorf("failed to get images: %v", err)
+		return fmt.Errorf("failed to check image: %v", err)
```
Comment (tariq1890) [inline pkg/provider/aws/dryrun.go:70]:
```suggestion
		return fmt.Errorf("failed to check image: %w", err)
```

File: pkg/provider/aws/image.go [hunk 23]
```diff
@@ -0,0 +1,155 @@
...
+		},
+		{
+			Name:   aws.String("owner-id"),
+			Values: awsOwner,
+		},
+	}
+
+	images, err := a.describeImages(filter)
+	if err != nil {
+		return fmt.Errorf("failed to describe images: %v", err)
```
Comment (tariq1890) [inline pkg/provider/aws/image.go:98]:
```suggestion
		return fmt.Errorf("failed to describe images: %w", err)
```

File: pkg/provider/aws/image.go [hunk 24]
```diff
@@ -0,0 +1,155 @@
...
+	var nextToken *string
+
+	for {
+		// Use the DescribeImages API to get a list of supported images in the current region
+		resp, err := a.ec2.DescribeImages(context.TODO(), &ec2.DescribeImagesInput{
+			NextToken: nextToken,
+			Filters:   filter,
+		})
+		if err != nil {
+			return images, fmt.Errorf("failed to describe images: %v", err)
```
Comment (tariq1890) [inline pkg/provider/aws/image.go:134]:
```suggestion
			return images, fmt.Errorf("failed to describe images: %w", err)
```

File: pkg/provider/aws/image.go [hunk 25]
```diff
@@ -0,0 +1,155 @@
...
+type ImageInfo struct {
+	ImageID      string
+	CreationDate string
+}
+
+type ByCreationDate []ImageInfo
+
+func (a ByCreationDate) Len() int           { return len(a) }
+func (a ByCreationDate) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a ByCreationDate) Less(i, j int) bool { return a[i].CreationDate > a[j].CreationDate }
```
Comment (tariq1890) [inline pkg/provider/aws/image.go:39]: For the sake of readability , I think the function should express a `<` rather than `>`

File: pkg/provider/aws/image.go [hunk 26]
```diff
@@ -0,0 +1,155 @@
...
+
+	images, err := a.describeImages(filter)
+	if err != nil {
+		return fmt.Errorf("failed to describe images: %v", err)
+	}
+
+	if len(images) == 0 {
+		return fmt.Errorf("no images found")
+	}
+	sort.Sort(ByCreationDate(images))
```
Comment (tariq1890) [inline pkg/provider/aws/image.go:104]: I feel like this is more readable and simpler as opposed to using a `Less` function implementation with inverted logic
```suggestion
	sort.Slice(images, func(i, j int) bool {
		return images[i].CreationDate > images[j].CreationDate
	})
```

Comment (tariq1890) [reply pkg/provider/aws/image.go:104]: The go devs also recommend it. See [here](https://pkg.go.dev/sort#example-package)

File: pkg/provider/aws/image.go [hunk 27]
```diff
@@ -0,0 +1,157 @@
...
+type ImageInfo struct {
+	ImageID      string
+	CreationDate string
+}
+
+type ByCreationDate []ImageInfo
+
+func (a ByCreationDate) Len() int           { return len(a) }
+func (a ByCreationDate) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a ByCreationDate) Less(i, j int) bool { return a[j].CreationDate < a[i].CreationDate }
```
Comment (tariq1890) [inline pkg/provider/aws/image.go:39]: This is counter-intuitive. I get that you want a descending order here, but the Less method should always represent a `i < j` equation. Instead of reversing the comparison here, we could just call a `Sort.Reverse()` at the time of invoking the Sort function
